validation/src/shared_table/mod.rs

```diff
@@ -318,59 +333,58 @@ pub struct PrimedParachainWork<Fetch, F> {
 	validate: F,
 }
 
-impl<Fetch, F, Err> Future for PrimedParachainWork<Fetch, F>
+impl<Fetch, F, Err> PrimedParachainWork<Fetch, F>
 	where
 		Fetch: Future<Item=PoVBlock,Error=Err>,
-		F: FnMut(&BlockId, &Collation) -> Result<OutgoingMessages, ()>,
+		F: FnMut(&BlockId, &PoVBlock, &CandidateReceipt) -> Result<(OutgoingMessages, ErasureChunk), ()>,
 		Err: From<::std::io::Error>,
 {
-	type Item = Validated;
-	type Error = Err;
-
-	fn poll(&mut self) -> Poll<Validated, Err> {
-		let work = &mut self.inner.work;
-		let candidate = &work.candidate_receipt;
+	pub async fn validate(mut self) -> Result<(Validated, Option<ErasureChunk>), Err> {
+		use futures03::compat::Future01CompatExt;
+		let candidate = &self.inner.work.candidate_receipt;
+		let pov_block = self.inner.work.fetch.compat().await?;
 
-		let pov_block = futures::try_ready!(work.fetch.poll());
 		let validation_res = (self.validate)(
 			&BlockId::hash(self.inner.relay_parent),
-			&Collation { pov: pov_block.clone(), receipt: candidate.clone() },
+			&pov_block,
+			&candidate,
 		);
 
 		let candidate_hash = candidate.hash();
 
 		debug!(target: "validation", "Making validity statement about candidate {}: is_good? {:?}",
 			candidate_hash, validation_res.is_ok());
 
-		let (validity_statement, result) = match validation_res {
-			Err(()) => (
-				GenericStatement::Invalid(candidate_hash),
-				Validation::Invalid(pov_block),
-			),
-			Ok(outgoing_targeted) => {
-				let outgoing_queues = crate::outgoing_queues(&outgoing_targeted)
-					.map(|(_target, root, data)| (root, data))
-					.collect();
-
-				self.inner.availability_store.make_available(Data {
-					relay_parent: self.inner.relay_parent,
-					parachain_id: work.candidate_receipt.parachain_index,
-					candidate_hash,
-					block_data: pov_block.block_data.clone(),
-					outgoing_queues: Some(outgoing_queues),
-				})?;
-
-				(
-					GenericStatement::Valid(candidate_hash),
-					Validation::Valid(pov_block, outgoing_targeted)
-				)
+		match validation_res {
+			Err(()) => Ok((
+				Validated {
+					statement: GenericStatement::Invalid(candidate_hash),
+					result: Validation::Invalid(pov_block),
+				},
+				None,
+			)),
+			Ok((outgoing_targeted, our_chunk)) => {
+				self.inner.availability_store.add_validator_index_and_n_validators(
```

Comment:
wasn't this done already in `get_or_instantiate`? The responsibility of setting this is unclear to me.